modules/channel-twilio/src/backend/twilio.ts

```diff
@@ -0,0 +1,236 @@
+
+    return entry.client
+  }
+
+  private async _handleIncomingMessage(req, res) {
+    const { botId } = req.params
+    const body = req.body
+
+    debugWebhook('_handleIncomingMessage:', { headers: req.headers, body: req.body })
+
+    res.status(200).send('<Response/>')
+
+    const fromNumber = body.To
+    const toNumber = body.From
+    this.fromToMap[toNumber] = { fromNumber, language: undefined }
+
+    if (undefined != body.Latitude) {
+      this.bp.logger.info('location')
+      await this._sendEvent(botId, toNumber, { text: body.Latitude + ',' + body.Longitude }, { type: 'location' })
+    } else if ('' != body.Body) {
+      this.bp.logger.info('message')
+      await this._sendEvent(botId, toNumber, { text: body.Body }, { type: 'text' })
+    }
+  }
+
+  private async _sendEvent(botId: string, senderId: string, message, args: { type: string }) {
+    args.type = message.text.toLowerCase() == 'reset' ? 'session_reset' : args.type || 'text'
+
+    await this.bp.events.sendEvent(
+      this.bp.IO.Event({
+        botId,
+        channel: 'twilio',
+        direction: 'incoming',
+        payload: message,
+        preview: message.text,
+        target: senderId,
+        ...args
+      })
+    )
+  }
+
+  private async _handleIncomingEvent(event: sdk.IO.IncomingEvent, next: sdk.IO.MiddlewareNextCallback) {
+    if (event.channel !== 'twilio') {
+      return next()
+    }
+
+    const langOptions = [event.nlu.language, event.nlu.detectedLanguage, event.state.user.language]
+
+    let lang = (langOptions.filter(l => l && l !== 'n/a')[0] || 'en').toLowerCase().substring(0, 2)
+
+    if (!entryPoint.translations[lang]) {
+      lang = 'en'
+    }
+
+    debugMessages(`language[${event.target}]=${lang} - options: `, langOptions)
+
+    this.fromToMap[event.target].language = lang
+
+    next(undefined, false)
+  }
+
+  private async _handleOutgoingEvent(event: sdk.IO.Event, next: sdk.IO.MiddlewareNextCallback) {
+    if (event.channel !== 'twilio') {
+      return next()
+    }
+
+    const messageType = event.type === 'default' ? 'text' : event.type
+    const recipient = this.fromToMap[event.target]
+    const tr = entryPoint.translations[recipient.language]
+    const twilio = this.getTwilioClientByBotId(event.botId)
+
+    if (messageType === 'text' || messageType === 'carousel' || messageType === 'dropdown') {
+      const elements = event.payload.elements || [event.payload]
+      for (const element of elements) {
+        let message = element.text || element.message
+        const options = element.quick_replies || element.options || element.buttons
+        if (messageType === 'carousel') {
+          message = element.title + (element.subtitle ? `\n${element.subtitle}` : '')
+          for (const option of options) {
+            if (option.type === 'open_url') {
+              message += `\n\n${option.title}: ${option.url}`
+              options.splice(options.indexOf(option), 1)
+            }
+          }
+        }
```

Comment:
This block is not very clear. Is it possible to extract them to individual methods?